File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
const fs = require("fs");

const reader = new class {
  data = fs
    .readFileSync(process.platform === "linux" ? 0 : "input.txt", "utf-8")
    .toString()
    .trim()
    .split("\n");
  
  cursor = 0;

  read() {
    return this.data[this.cursor++];
  }

  readIntArray() {
    return this.read().split(" ").map(s => parseInt(s));
  }
}

function solve() {
  const [N, M, source] = reader.readIntArray();

  const graph = Array.from({ length: N + 1 }, () => []);
  let visited = new Array(N + 1).fill(false);
  let order = "";

  for (let _ = 0; _ < M; _++) {
    const [start, end] = reader.readIntArray();
    graph[start].push(end);
    graph[end].push(start);
  }

  for (let i = 1; i <= N; i++) {
    graph[i].sort((a, b) => a - b);
  }

  const clearVisited = () => {
    visited = visited.fill(false);
  };

  const dfs = (now) => {
    order += `${now} `;
    visited[now] = true;
    for (const next of graph[now]) {
      if (!visited[next]) {
        dfs(next);
      }
    }
  };
  
  const bfs = (source) => {
    const queue = [];
    queue.push(source);
    visited[source] = true;
    
    while (queue.length > 0) {
      const now = queue.shift();
      order += `${now} `;
      for (const next of graph[now]) {
        if (!visited[next]) {
          visited[next] = true;
          queue.push(next);
        }
      }
    }
  };
  
  dfs(source);

  order += "\n";
  clearVisited();
  
  bfs(source);

  return order;
}

console.log(solve());

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

input = sys.stdin.readline
print = sys.stdout.write

nr_vertices, nr_edges, entry_vertex = map(int, input().split())

visited = [False] * (nr_vertices + 1)
graph = [[] for _ in range(nr_vertices + 1)]

for _ in range(nr_edges):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

for i in range(1, nr_vertices + 1):
    graph[i].sort()

def dfs(curr_vertex):
    for next_vertex in graph[curr_vertex]:
        if not visited[next_vertex]:
            visited[next_vertex] = True
            print(f"{next_vertex} ")
            dfs(next_vertex)

visited[entry_vertex] = True
print(f"{entry_vertex} ")
dfs(entry_vertex)

print("\n")
visited = [False] * (nr_vertices + 1)

def bfs():
    queue = deque()
    queue.append(entry_vertex)
    visited[entry_vertex] = True

    while len(queue) > 0:
        curr_vertex = queue.popleft()
        print(f"{curr_vertex} ")

        for next_vertex in graph[curr_vertex]:
            if not visited[next_vertex]:
                visited[next_vertex] = True
                queue.append(next_vertex)

bfs()
